src/crypto/muhash.cpp

```diff
@@ -141,98 +123,386 @@ void Num3072::FullReduce()
     }
 }
 
-Num3072 Num3072::GetInverse() const
+namespace {
+/** A type representing a number in signed limb representation. */
+struct Num3072Signed
 {
-    // For fast exponentiation a sliding window exponentiation with repunit
-    // precomputation is utilized. See "Fast Point Decompression for Standard
-    // Elliptic Curves" (Brumley, Järvinen, 2008).
+    /** The represented value is sum(limbs[i]*2^(SIGNED_LIMB_SIZE*i), i=0..SIGNED_LIMBS-1).
+     *  Note that limbs may be negative, or exceed 2^SIGNED_LIMB_SIZE-1. */
+    signed_limb_t limbs[SIGNED_LIMBS];
+
+    /** Construct a Num3072Signed with value 0. */
+    Num3072Signed()
+    {
+        memset(limbs, 0, sizeof(limbs));
+    }
 
-    Num3072 p[12]; // p[i] = a^(2^(2^i)-1)
-    Num3072 out;
+    /** Convert a Num3072 to a Num3072Signed. Output will be normalized and in
+     *  range 0..2^3072-1. */
+    void FromNum3072(const Num3072& in)
+    {
+        double_limb_t c = 0;
+        int b = 0, outpos = 0;
+        for (int i = 0; i < LIMBS; ++i) {
+            c += ((double_limb_t)in.limbs[i]) << b;
```

Comment:
please avoid c-style cast, use c++11 style functional casts

`double_limb_t(in.limbs[i])`